registry/api/v2/urls.go

```diff
@@ -83,22 +88,25 @@ func NewURLBuilderFromRequest(r *http.Request, relative bool) *URLBuilder {
 		host = h
 	}
 
-	portLessHost, port := host, ""
-	if !isIPv6Address(portLessHost) {
+	portlessHost, port := host, ""
+	if !isIPv6Address(host) {
 		// with go 1.6, this would treat the last part of IPv6 address as a port
-		portLessHost, port, _ = net.SplitHostPort(host)
+		portlessHost, port, _ = net.SplitHostPort(host)
+		if len(portlessHost) > 0 {
+			host = portlessHost
+		}
 	}
 	if forwardedPort := r.Header.Get("X-Forwarded-Port"); len(port) == 0 && len(forwardedPort) > 0 {
 		ports := strings.SplitN(forwardedPort, ",", 2)
 		forwardedPort = strings.TrimSpace(ports[0])
 		if _, err := strconv.ParseInt(forwardedPort, 10, 32); err == nil {
 			port = forwardedPort
+			if port == portMap[scheme] {
```

Comment:
Would change the logic here to...

```
if forwardedPort != portMap[scheme] {
	port = forwardedPort
}
```

This would prevent altering the `port` value set earlier. Can you also add a test to make sure that a port set in one of the host fields will not get overridden by this.